pkg/handlers/archive.go

```diff
@@ -183,29 +185,25 @@ func (a *Archive) ReadToMax(ctx context.Context, reader io.Reader) (data []byte,
 			logger.Error(err, "Panic occurred when reading archive")
 		}
 	}()
-	fileContent := bytes.Buffer{}
-	logger.V(5).Info("Remaining buffer capacity", "bytes", maxSize-a.size)
-	for i := 0; i <= maxSize/512; i++ {
-		if common.IsDone(ctx) {
-			return nil, ctx.Err()
-		}
-		fileChunk := make([]byte, 512)
-		bRead, err := reader.Read(fileChunk)
-		if err != nil && !errors.Is(err, io.EOF) {
-			return []byte{}, err
-		}
-		a.size += bRead
-		if len(fileChunk) > 0 {
-			fileContent.Write(fileChunk[0:bRead])
-		}
-		if bRead < 512 {
-			return fileContent.Bytes(), nil
-		}
-		if a.size >= maxSize && bRead == 512 {
-			logger.V(2).Info("Max archive size reached.")
-			return fileContent.Bytes(), nil
-		}
+
+	if common.IsDone(ctx) {
+		return nil, ctx.Err()
+	}
+
+	var fileContent bytes.Buffer
+	// Using io.CopyBuffer for performance advantages. Though buf is mandatory
+	// for the method, due to the internal implementation of io.CopyBuffer, when
+	// *bytes.Buffer implements io.WriterTo or io.ReaderFrom, the provided buf
+	// is simply ignored. Thus, we can pass nil for the buf parameter.
+	_, err = io.CopyBuffer(&fileContent, reader, nil)
+	if err != nil && !errors.Is(err, io.EOF) {
+		return nil, err
+	}
+	if fileContent.Len() >= maxSize {
+		logger.V(2).Info("Max archive size reached.")
+		return fileContent.Bytes()[:maxSize], nil
 	}
```

Comment:
fixed w/ limitReader i think.

Reply:
updated benchmark screenshot as well.

Reply:
We don't get the super blazing fast copy behavior b/c LimitReader doesn't not have a WriteTo method like a bytes.Reader 😢 . We have to pay the penalty of 10 more allocs.

Reply:
Ah, that's unfortunate. Should we provide a `buf` in that case?

Reply:
It would still be ignored because the dst implements ReadFrom, but we still need to provide it.
```
// If either src implements WriterTo or dst implements ReaderFrom,
// buf will not be used to perform the copy.
```